_posts/2018-11-11-fs.md

```diff
@@ -0,0 +1,81 @@
+  char buf2[113];
+  memset(buf2, 0, sizeof(buf2));
+  listxattr(xattr, buf2, sizeof(buf2));
+  removexattr(xattr, "user.mime_type");
+}
+```
+The buffer overflow happens in the kernel's `fs/ext4/xattr.c:ext4_xattr_set_entry() 
+` function, because of a [missed check](https://elixir.bootlin.com/linux/v4.15/source/fs/ext4/xattr.c#L1671) on the `size` parameter of the [`memmove()`](http://man7.org/linux/man-pages/man3/memmove.3.html) function, which is negative when trying to remove an extended attribute (`user.mime_type`) for the file `foo/bar/xattr` in the crafted image. 
+The fix is fairly straightforward in this case--adding a [value check](https://bugzilla.kernel.org/attachment.cgi?id=276147&action=diff).
+
+
+
+### Nabla and GVisor
+Nabla's [philosophy](https://nabla-containers.github.io/) of limiting access to the host kernel means that a container's file access does not interface with the host's filesystem.
+Specifically, Nabla does not use a 9p or a passthrough protocol, with syscalls on the filesystem handled at the LibOS (rumprun) level.
+Thus, this bug would never get triggered<sup>3</sup>, even if the host kernel is vulnerable, since `removexattr()` is not propagated to the host. 
+Nabla's minimal *touch points* into the host can be seen [here](https://github.com/nabla-containers/solo5/blob/ukvm-linux-seccomp/ukvm/ukvm_hv_linux.c#L53-L106).
+
+The bug does not surface in case of Gvisor as well.
+Neither [Sentry](https://github.com/google/gvisor/blob/master/pkg/sentry/syscalls/linux/linux64.go#L239-L250) nor [Gofer](https://github.com/google/gvisor/blob/master/runsc/fsgofer/filter/config.go) support the syscall--"`Removexattr, requires filesystem support`"
+
+
+
+### So what?
+This particular bug doesn't present a serious security vulnerability in a container setting, since:
+1. it needs a crafted image to be exposed as rootfs or a volume to the containers,
+2. the end product of a successful exploit means only a killed process within the adversary container, and
+3. the kernel remains functional despite the bug being triggered.
+
+But it does serve as an example of what's possible when various *touch points* into the host kernel are available, and the need to minimize them. 
+Even from a filesystem *touch-point* perspective, we cannot eliminate the existence of [bugs](https://groups.google.com/forum/#!msg/syzkaller/_8MZkKL2-QU/PA0q5XULEAAJ) that do not require a crafted image, as well as the potential of [bugs](https://bugzilla.kernel.org/show_bug.cgi?id=96971) causing a full kernel panic, or serious side-effects such as [privilege escalation](http://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2015-1328).
+
+In this blog, although we chose the filesystem interface to highlight Nabla's *touch-point* (attack surface) minimization philosophy, the same is applicable to other components as well, including bypassing [bugs](https://www.cvedetails.com/vulnerability-list/vendor_id-7506/Qemu.html) in the virtualization interface itself!
```

Comment:
might be worth adding a reference to our hotcloud '18 paper "Say Goodbye to Virtualization for a Safer Cloud"